rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,650 @@
+         - (-) Have to rename all files in a project to change the version
+           - This is somewhat worse than replacing a magic string which is fixed to the beginning of the file
+     - (-) Makes filenames longer, introduces visual noise
+         - This is the cost of being explicit
+     - (-) Enforces narrow restrictions on what information can be encoded and how
+       - The only reasonable alternative is `-`, e.g. `default.7-nix`
+         - `.`
+             - (-) Nixpkgs has been packaging Linux kernels as `linux-${major}.${minor}.nix`
+               - This may break backwards compatibility of newer evaluators with existing code in surprising ways
+         - No separator
+             – (-) Hard to discern visually
+         - `-`
+             - (+) Visually not intrusive
+         - `_`
+             – (-) Visually more intrusive
+         - `^`
+             – (-) Overlaps with derivation output syntax
+         - All of the following characters will interfere with some tooling:
+             - `!` - shells
+             - `"` - shells
+             - `#` - URLs
+             - `$` - shells
+             - `%` - URLs
+             - `&` - shells, URLs
+             - `+` - URLs
+             - `,` - natural language
+             - `/` - paths, URLs
+             - `:` - URLs
+             - `;` - shells
+             - `=` - URLs, Nix language
+             - `?` - URLs
+             - `@` - URLs, Nix language
+             - `\` - Windows paths
+     - (-) `default.nix` resolving needs specification:
+         - If for `import ./foo`, all of `./foo/default.nix{6,7,8}` exist, pick the one matching the version used by the evaluator, otherwise fail
+             - Then you'd have to specify a file using a different version explicitly
+
+   - Language versioning per "project" in a sidecar file
+       - (+) This would easily allow inheriting the language version across imports (obviating many specifications in this proposal)
+       - (-) There is currently no notion of "project" in the Nix language
+         - (-) Attempting to establish one would be a large undertaking and not immediately help solving the problem at hand
+       - (-) Cannot be introduced gradually, particularly relevant for a large codebase like Nixpkgs
+       - (-) It would require a separate language to encode project metadata such as the language version
+           - The `edition` field was [removed from the flakes schema](https://github.com/NixOS/nix/commit/e5ea01c1a8bbd328dcc576928bf3e4271cb55399) for that reason, as it not not allow distinguishing data from metadata
+           - (+) Other languages do the same (Python, Haskell, Rust, JavaScript, ...)
+             - (-) Recursive (albeit smaller) problem of managing the additional language for project metadata
+
+   </details>
+
+1. The following syntax is used for declaring the language version of a Nix expression:
+
+   ```
+   version \d*;
+   ```
```

Comment:
This is problematic because you'll either have to:
- Reserve `version` as a special keyword, which would breaking compatibility itself because `version` can't be used as an identifier anymore.
- Change the parser to be able to detect whether `version` is used as a keyword or an identifier. I'm not sure if that's doable, requires arbitrary look-ahead and might be very slow

Reply:
It's not clear if this is restricted to the first token in the file. If so it may be acceptable. While technically `someVariable` is a valid nix expression it will not resolve in a file unless it is a builtin global. I agree that we can't make it a keyword in general, but we can likely make it a keyword in a specific context.

Reply:
@kevincox Ah forgot about that, yeah I think that's what's specified here, in which case this should work just fine.

Reply:
I would definitely like to see some more clarification and ideally an explanation of why this is backwards-compatible.